File: 113. Path Sum II/113. Path Sum II.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def pathSum(self, root, targetSum):
        """
        :type root: Optional[TreeNode]
        :type targetSum: int
        :rtype: List[List[int]]
        """
        def findPaths(node, target, path, result):
            if not node:
                return
            
            path.append(node.val)
            
            if not node.left and not node.right and target == node.val:
                result.append(path[:])
            
            findPaths(node.left, target - node.val, path, result)
            findPaths(node.right, target - node.val, path, result)
            
            path.pop()
        
        result = []
        findPaths(root, targetSum, [], result)
        return result

File: 113. Path Sum II/113. Path Sum II.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
    final List<List<int>> result = [];
    
    void findPaths(TreeNode? root, int targetSum, List<int> path) {
        if (root == null) return;
        
        path.add(root.val);
        
        if (root.left == null && root.right == null && targetSum == root.val) {
            result.add(List.from(path));
        }
        
        findPaths(root.left, targetSum - root.val, path);
        findPaths(root.right, targetSum - root.val, path);
        
        path.removeLast();
    }
    
    List<List<int>> pathSum(TreeNode? root, int targetSum) {
        result.clear();
        findPaths(root, targetSum, []);
        return result;
    }
}